File: js_example/src/substrate_client.js
import { ApiPromise, WsProvider } from '@polkadot/api';

function toBytesArg(value) {
  if (Array.isArray(value)) return value;
  return Array.from(value);
}

function callArgNames(api, section, method) {
  const call = api.tx?.[section]?.[method];
  return call?.meta?.args?.map((arg) => arg.name.toString()) ?? [];
}

function normalizeName(name) {
  return name.replace(/_/g, '').toLowerCase();
}

export function didSupportsKeyMaterial(api) {
  return callArgNames(api, 'did', 'addKey').some((name) => normalizeName(name) === 'keymaterial');
}

export function materialForRuntime(api, keyMaterial) {
  if (didSupportsKeyMaterial(api)) return keyMaterial;
  if (keyMaterial?.Multikey) return keyMaterial.Multikey;
  throw new Error('Connected runtime does not support JWK key material');
}

export async function createApi(url) {
  process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';
  const provider = new WsProvider(url);
  const api = await ApiPromise.create({ provider });
  return api;
}

export async function getFreeBalance(api, address) {
  const { data } = await api.query.system.account(address);
  return data.free.toString();
}

async function submitExtrinsic(tx, account) {
  return new Promise((resolve, reject) => {
    let unsub = null;
    tx.signAndSend(account, (result) => {
      const done = result.status?.isInBlock || result.status?.isFinalized;
      if (done || result.dispatchError) {
        if (unsub) unsub();
        resolve(result);
      }
    })
      .then((u) => {
        unsub = u;
      })
      .catch(reject);
  });
}

export async function createDid(api, account, publicKey, didSignature) {
  const tx = api.tx.did.createDid(toBytesArg(publicKey), toBytesArg(didSignature));
  return submitExtrinsic(tx, account);
}

export async function addKey(
  api,
  account,
  didId,
  keyIdSuffix,
  keyMaterial,
  roles,
  controller,
  didSignature
) {
  const material = materialForRuntime(api, keyMaterial);
  const tx = api.tx.did.addKey(
    toBytesArg(didId),
    keyIdSuffix == null ? null : toBytesArg(keyIdSuffix),
    material,
    roles,
    controller == null ? null : toBytesArg(controller),
    toBytesArg(didSignature)
  );
  return submitExtrinsic(tx, account);
}

export async function revokeKey(api, account, didId, keyId, didSignature) {
  const tx = api.tx.did.revokeKey(
    toBytesArg(didId),
    toBytesArg(keyId),
    toBytesArg(didSignature)
  );
  return submitExtrinsic(tx, account);
}

export async function deactivateDid(api, account, didId, didSignature) {
  const tx = api.tx.did.deactivateDid(toBytesArg(didId), toBytesArg(didSignature));
  return submitExtrinsic(tx, account);
}

export async function addService(
  api,
  account,
  didId,
  serviceId,
  serviceType,
  endpoint,
  didSignature
) {
  const service = {
    id: toBytesArg(serviceId),
    service_type: toBytesArg(serviceType),
    endpoint: toBytesArg(endpoint),
  };
  const tx = api.tx.did.addService(toBytesArg(didId), service, toBytesArg(didSignature));
  return submitExtrinsic(tx, account);
}

export async function removeService(api, account, didId, serviceId, didSignature) {
  const tx = api.tx.did.removeService(
    toBytesArg(didId),
    toBytesArg(serviceId),
    toBytesArg(didSignature)
  );
  return submitExtrinsic(tx, account);
}

export async function setMetadata(api, account, didId, key, value, didSignature) {
  const entry = {
    key: toBytesArg(key),
    value: toBytesArg(value),
  };
  const tx = api.tx.did.setMetadata(toBytesArg(didId), entry, toBytesArg(didSignature));
  return submitExtrinsic(tx, account);
}

export async function removeMetadata(api, account, didId, key, didSignature) {
  const tx = api.tx.did.removeMetadata(
    toBytesArg(didId),
    toBytesArg(key),
    toBytesArg(didSignature)
  );
  return submitExtrinsic(tx, account);
}

export async function rotateKey(
  api,
  account,
  didId,
  oldKeyId,
  newKeyMaterial,
  newKeyIdSuffix,
  newController,
  roles,
  didSignature
) {
  const material = materialForRuntime(api, newKeyMaterial);
  const tx = api.tx.did.rotateKey(
    toBytesArg(didId),
    toBytesArg(oldKeyId),
    material,
    newKeyIdSuffix == null ? null : toBytesArg(newKeyIdSuffix),
    newController == null ? null : toBytesArg(newController),
    roles,
    toBytesArg(didSignature)
  );
  return submitExtrinsic(tx, account);
}

export async function updateRoles(api, account, didId, keyId, roles, didSignature) {
  const tx = api.tx.did.updateRoles(
    toBytesArg(didId),
    toBytesArg(keyId),
    roles,
    toBytesArg(didSignature)
  );
  return submitExtrinsic(tx, account);
}

export async function registerSchema(
  api,
  account,
  schemaJson,
  schemaUri,
  issuerDid,
  didSignature
) {
  const tx = api.tx.schema.registerSchema(
    toBytesArg(schemaJson),
    toBytesArg(schemaUri),
    toBytesArg(issuerDid),
    toBytesArg(didSignature)
  );
  return submitExtrinsic(tx, account);
}

export async function deprecateSchema(api, account, schemaId, issuerDid, didSignature) {
  const tx = api.tx.schema.deprecateSchema(
    toBytesArg(schemaId),
    toBytesArg(issuerDid),
    toBytesArg(didSignature)
  );
  return submitExtrinsic(tx, account);
}

File: python_example/src/app/substrate_client.py
import ssl

from substrateinterface import SubstrateInterface


def create_substrate(url: str) -> SubstrateInterface:
    return SubstrateInterface(
        url=url,
        ws_options={"sslopt": {"cert_reqs": ssl.CERT_NONE}},
    )


def get_free_balance(substrate: SubstrateInterface, address: str) -> int:
    account_info = substrate.query("System", "Account", [address])
    return account_info.value["data"]["free"]


def _call_field_names(substrate: SubstrateInterface, call_function: str) -> list[str]:
    metadata = substrate.get_metadata_call_function("Did", call_function)
    return [field["name"] for field in metadata["fields"]]


def did_supports_key_material(substrate: SubstrateInterface) -> bool:
    return "key_material" in _call_field_names(substrate, "add_key")


def _material_to_legacy_public_key(key_material: dict) -> bytes:
    if "Multikey" in key_material:
        return key_material["Multikey"]
    raise ValueError("Connected runtime does not support JWK key material")


def create_did(substrate: SubstrateInterface, account, public_key: bytes, did_signature: bytes):
    call = substrate.compose_call(
        call_module="Did",
        call_function="create_did",
        call_params={
            "public_key": public_key,
            "did_signature": did_signature,
        },
    )
    extrinsic = substrate.create_signed_extrinsic(call=call, keypair=account)
    return substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)


def add_key(
    substrate: SubstrateInterface,
    account,
    did_id: bytes,
    key_id_suffix: bytes | None,
    key_material: dict,
    roles: list[str],
    controller: bytes | None,
    did_signature: bytes,
):
    supports_key_material = did_supports_key_material(substrate)
    material_param = (
        {"key_material": key_material}
        if supports_key_material
        else {"public_key": _material_to_legacy_public_key(key_material)}
    )
    call = substrate.compose_call(
        call_module="Did",
        call_function="add_key",
        call_params={
            "did_id": did_id,
            "key_id_suffix": key_id_suffix,
            **material_param,
            "roles": roles,
            "controller": controller,
            "did_signature": did_signature,
        },
    )
    extrinsic = substrate.create_signed_extrinsic(call=call, keypair=account)
    return substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)


def revoke_key(
    substrate: SubstrateInterface,
    account,
    did_id: bytes,
    key_id: bytes,
    did_signature: bytes,
):
    call = substrate.compose_call(
        call_module="Did",
        call_function="revoke_key",
        call_params={
            "did_id": did_id,
            "key_id": key_id,
            "did_signature": did_signature,
        },
    )
    extrinsic = substrate.create_signed_extrinsic(call=call, keypair=account)
    return substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)


def deactivate_did(
    substrate: SubstrateInterface,
    account,
    did_id: bytes,
    did_signature: bytes,
):
    call = substrate.compose_call(
        call_module="Did",
        call_function="deactivate_did",
        call_params={
            "did_id": did_id,
            "did_signature": did_signature,
        },
    )
    extrinsic = substrate.create_signed_extrinsic(call=call, keypair=account)
    return substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)


def register_schema(
    substrate: SubstrateInterface,
    account,
    schema_json: bytes,
    schema_uri: bytes,
    issuer_did: bytes,
    did_signature: bytes,
):
    call = substrate.compose_call(
        call_module="Schema",
        call_function="register_schema",
        call_params={
            "schema_json": schema_json,
            "schema_uri": schema_uri,
            "issuer_did": issuer_did,
            "did_signature": did_signature,
        },
    )
    extrinsic = substrate.create_signed_extrinsic(call=call, keypair=account)
    return substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)


def deprecate_schema(
    substrate: SubstrateInterface,
    account,
    schema_id: bytes,
    issuer_did: bytes,
    did_signature: bytes,
):
    call = substrate.compose_call(
        call_module="Schema",
        call_function="deprecate_schema",
        call_params={
            "schema_id": schema_id,
            "issuer_did": issuer_did,
            "did_signature": did_signature,
        },
    )
    extrinsic = substrate.create_signed_extrinsic(call=call, keypair=account)
    return substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)


def add_service(
    substrate: SubstrateInterface,
    account,
    did_id: bytes,
    service_id: bytes,
    service_type: bytes,
    endpoint: bytes,
    did_signature: bytes,
):
    call = substrate.compose_call(
        call_module="Did",
        call_function="add_service",
        call_params={
            "did_id": did_id,
            "service": {
                "id": service_id,
                "service_type": service_type,
                "endpoint": endpoint,
            },
            "did_signature": did_signature,
        },
    )
    extrinsic = substrate.create_signed_extrinsic(call=call, keypair=account)
    return substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)


def remove_service(
    substrate: SubstrateInterface,
    account,
    did_id: bytes,
    service_id: bytes,
    did_signature: bytes,
):
    call = substrate.compose_call(
        call_module="Did",
        call_function="remove_service",
        call_params={
            "did_id": did_id,
            "service_id": service_id,
            "did_signature": did_signature,
        },
    )
    extrinsic = substrate.create_signed_extrinsic(call=call, keypair=account)
    return substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)


def set_metadata(
    substrate: SubstrateInterface,
    account,
    did_id: bytes,
    key: bytes,
    value: bytes,
    did_signature: bytes,
):
    call = substrate.compose_call(
        call_module="Did",
        call_function="set_metadata",
        call_params={
            "did_id": did_id,
            "entry": {
                "key": key,
                "value": value,
            },
            "did_signature": did_signature,
        },
    )
    extrinsic = substrate.create_signed_extrinsic(call=call, keypair=account)
    return substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)


def remove_metadata(
    substrate: SubstrateInterface,
    account,
    did_id: bytes,
    key: bytes,
    did_signature: bytes,
):
    call = substrate.compose_call(
        call_module="Did",
        call_function="remove_metadata",
        call_params={
            "did_id": did_id,
            "key": key,
            "did_signature": did_signature,
        },
    )
    extrinsic = substrate.create_signed_extrinsic(call=call, keypair=account)
    return substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)


def rotate_key(
    substrate: SubstrateInterface,
    account,
    did_id: bytes,
    old_key_id: bytes,
    new_key_material: dict,
    new_key_id_suffix: bytes | None,
    new_controller: bytes | None,
    roles: list[str],
    did_signature: bytes,
):
    supports_key_material = did_supports_key_material(substrate)
    material_param = (
        {"new_key_material": new_key_material}
        if supports_key_material
        else {"new_public_key": _material_to_legacy_public_key(new_key_material)}
    )
    call = substrate.compose_call(
        call_module="Did",
        call_function="rotate_key",
        call_params={
            "did_id": did_id,
            "old_key_id": old_key_id,
            **material_param,
            "new_key_id_suffix": new_key_id_suffix,
            "new_controller": new_controller,
            "roles": roles,
            "did_signature": did_signature,
        },
    )
    extrinsic = substrate.create_signed_extrinsic(call=call, keypair=account)
    return substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)


def update_roles(
    substrate: SubstrateInterface,
    account,
    did_id: bytes,
    key_id: bytes,
    roles: list[str],
    did_signature: bytes,
):
    call = substrate.compose_call(
        call_module="Did",
        call_function="update_roles",
        call_params={
            "did_id": did_id,
            "key_id": key_id,
            "roles": roles,
            "did_signature": did_signature,
        },
    )
    extrinsic = substrate.create_signed_extrinsic(call=call, keypair=account)
    return substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)
